cmd/pdatagen/internal/base_structs.go

```diff
@@ -338,19 +338,91 @@ func Test${structName}_CopyTo(t *testing.T) {
 	assert.EqualValues(t, generateTest${structName}(), ms)
 }`
 
-const messageGenerateTestTemplate = `func generateTest${structName}() ${structName} {
+const messagePtrGenerateTestTemplate = `func generateTest${structName}() ${structName} {
 	tv := New${structName}()
 	tv.InitEmpty()
 	fillTest${structName}(tv)
 	return tv
 }`
 
-const messageFillTestHeaderTemplate = `func fillTest${structName}(tv ${structName}) {`
-const messageFillTestFooterTemplate = `}`
+const messagePtrFillTestHeaderTemplate = `func fillTest${structName}(tv ${structName}) {`
+const messagePtrFillTestFooterTemplate = `}`
+
+const messageValueTemplate = `${description}
+//
+// This is a reference type, if passed by value and callee modifies it the
+// caller will see the modification.
+//
+// Must use New${structName} function to create new instances.
+// Important: zero-initialized instance is not valid for use.
+type ${structName} struct {
+	// orig points to the pointer ${originName} field contained somewhere else.
+	orig *${originName}
+}
+
+var emptyValue${structName} = ${originName}{}
+
+func new${structName}(orig *${originName}) ${structName} {
+	return ${structName}{orig: orig}
+}
+
+// New${structName} creates a new empty ${structName}.
+//
+// This must be used only in testing code since no "Set" method available.
+func New${structName}() ${structName} {
+	return new${structName}(&${originName}{})
+}
+
+// Deprecated: This function will be removed soon.
+func (ms ${structName}) InitEmpty() {
+	*ms.orig = ${originName}{}
+}
+
+// Deprecated: Use IsEmpty instead
+func (ms ${structName}) IsNil() bool {
+	return ms.IsEmpty()
+}`
+
+const messageValueIsEmptyHeaderTemplate = `// IsEmpty returns true if the underlying data are equivalent with an empty message.
+func (ms ${structName}) IsEmpty() bool {
```

Comment:
I wonder if we can completely eliminate IsEmpty for value types. All the usages in the codebase that I can see I think can are very likely possible to remove safely.

Reply:
Done.